pytential/symbolic/execution.py

```diff
@@ -127,27 +127,27 @@ def _map_elementwise_reduction(self, reduction_name, expr):
         from arraycontext import make_loopy_program
         from meshmode.transform_metadata import (
                 ConcurrentElementInameTag, ConcurrentDOFInameTag)
+        actx = self.array_context
 
-        @memoize_in(self.places, "elementwise_node_"+reduction_name)
+        @memoize_in(actx, f"elementwise_node_{reduction_name}")
         def node_knl():
             t_unit = make_loopy_program(
                     """{[iel, idof, jdof]:
                         0<=iel<nelements and
                         0<=idof, jdof<ndofs}""",
                     """
-                    result[iel, idof] = %s(jdof, operand[iel, jdof])
+                    <> el_result = %s(jdof, operand[iel, jdof])
+                    result[iel, idof] = el_result
                     """ % reduction_name,
-                    name="nodewise_reduce")
+                    name=f"elementwise_node_{reduction_name}")
 
             return lp.tag_inames(t_unit, {
                 "iel": ConcurrentElementInameTag(),
                 "idof": ConcurrentDOFInameTag(),
                 })
 
-        @memoize_in(self.places, "elementwise_"+reduction_name)
+        @memoize_in(actx, f"elementwise_element_{reduction_name}")
         def element_knl():
-            # FIXME: This computes the reduction value redundantly for each
-            # output DOF.
```

Comment:
I tried fixing that `FIXME` by storing the result of the elementwise reduction. Does that look better?

Reply:
Doesn't seem like this should be `ConcurrentDOFInameTag` anymore for `idof`?

Reply:
Doesn't look like it is?

Reply:
The `idof` from the `elementwise_node` kernel does have that tag (on line 146).

Reply:
Apparently none of the tests were actually using this with `GRANULARITY_NODE`! I added one in https://github.com/inducer/pytential/compare/8ab424c713e31b03cd30864c2b479a8357263a83..fd3e217d830eda7cbb991c7278db07a6c3f05475.

This was failing with 
```
loopy.diagnostic.LoopyError: instruction 'insn_jdof_init' does not use all local hw axes (available: 0 used: ). Calling loopy.add_inames_for_unused_hw_axes(...) might help.
```
when `"idof": ConcurrentDOFInameTag()` was in the tags. Not sure if the best solution is to remove that, but it does feel like some sort of barrier / sync is needed to compute `el_result` before broadcasting it to all the `idofs`.

Reply:
I pushed a revert of this before changing my mind. It's a bit of a CPU vs GPU conundrum, but since pytential is still a predominantly-CPU code, the lack of parallelism is not a tremendous concern.